src/api/transportTypes.js

```diff
@@ -31,25 +31,47 @@ export type ApiResponseSuccess = {|
   msg: '',
 |};
 
-/** List of error codes at https://github.com/zulip/zulip/blob/master/zerver/lib/exceptions.py */
-export type ApiErrorCode =
-  | 'BAD_REQUEST'
-  | 'REQUEST_VARIABLE_MISSING'
-  | 'REQUEST_VARIABLE_INVALID'
-  | 'BAD_IMAGE'
-  | 'REALM_UPLOAD_QUOTA'
-  | 'BAD_NARROW'
-  | 'MISSING_HTTP_EVENT_HEADER'
-  | 'STREAM_DOES_NOT_EXIST'
-  | 'UNAUTHORIZED_PRINCIPAL'
-  | 'UNEXPECTED_WEBHOOK_EVENT_TYPE'
-  | 'BAD_EVENT_QUEUE_ID'
-  | 'CSRF_FAILED'
-  | 'INVITATION_FAILED'
-  | 'INVALID_ZULIP_SERVER';
+/**
+ * A list of current error codes can be found at:
+ *   https://github.com/zulip/zulip/blob/master/zerver/lib/exceptions.py
+ *
+ * Unfortunately, the `code` entry is a relatively late addition to the Zulip
+ * API: it hasn't always been present in responses, and even on current versions
+ * of the server, may still not be present in some. Since we also can't reliably
+ * synthesize it locally, we can't guarantee that a `code` property will be
+ * present in our data structures, especially not when connecting to older
+ * versions of the Zulip server.
+ *
+ * It would still be possible to make ApiErrorCode an enumerated type, if not
+ * for the dual problem: when connecting to _newer_ Zulip servers, we may see
+ * values of `code` not known to this version of the client!
+ *
+ * (Note, in particular, that "newer" doesn't necessarily mean "newer than this
+ * client", but "newer than the last time someone updated the error code list
+ * from the server". We have no mechanism in place to share the set of error
+ * codes... and, given the preceding facts, it is not clear that the existence
+ * of such a mechanism would be helpful.)
+ *
+ * For all that, `code` is not completely useless. It _is_ guaranteed to be
+ * returned, and stable, for certain known classes of errors arising from
+ * certain known endpoints, and all new errors should have a `code` value.
```

Comment:
(same discussion as above)

Oh just for precision in talking about it, though: I think in fact all errors from non-ancient servers (since 1.7.0 -- specifically 709c3b50f and near ancestors) do have *a* `code`; it's just that for most of them, it's a generic  'BAD_REQUEST'.

Reply:
That _completely_ changes my preferred approach here. If the server's filling in `BAD_REQUEST` for "code not otherwise specified", then we might as well do the same on our end for pre-1.7.0 servers, and just document that it's an explicit catch-all.

Reply:
Huh -- yeah, I definitely agree! Had not put the pieces together to think of that. :slightly_smiling_face: That will simplify some things.